async_upnp_client/profiles/profile.py

```diff
@@ -179,53 +187,156 @@ def _interesting_service(self, service: UpnpService) -> bool:
                 continue
 
-            on_event: EventCallbackType = self._on_event
-            service.on_event = on_event
-            if self._event_handler.sid_for_service(service) is None:
-                _LOGGER.debug("Subscribing to service: %s", service)
-                success, _, timeout = await self._event_handler.async_subscribe(
-                    service, timeout=SUBSCRIBE_TIMEOUT
+            try:
+                new_sid, timeout = await self._event_handler.async_resubscribe(
+                    sid, timeout=SUBSCRIBE_TIMEOUT
                 )
-                if not success:
-                    _LOGGER.debug("Failed subscribing to: %s", service)
-                elif timeout:
-                    timeouts.append(timeout)
+            except UpnpError as err:
+                if isinstance(err, UpnpConnectionError):
+                    # Device has gone offline
+                    self.device.available = False
+                _LOGGER.error("Failed (re-)subscribing to: %s, reason: %s", sid, err)
+                if notify_errors:
+                    # Notify event listeners that something has changed
+                    self._on_event(service, [])
+                else:
+                    raise
             else:
-                _LOGGER.debug("Resubscribing to service: %s", service)
-                success, _, timeout = await self._event_handler.async_resubscribe(
-                    service, timeout=SUBSCRIBE_TIMEOUT
-                )
-
-                # could not renew subscription, try subscribing again
-                if not success:
-                    _LOGGER.debug("Failed resubscribing to: %s", service)
+                self._subscriptions[new_sid] = now + timeout.total_seconds()
+
+    async def _resubscribe_loop(self) -> None:
+        """Periodically resubscribes to current subscriptions."""
+        _LOGGER.debug("_resubscribe_loop started")
+        while self._subscriptions:
+            next_renewal = min(self._subscriptions.values())
+            wait_time = next_renewal - time.monotonic() - RESUBSCRIBE_TOLERANCE_SECS
+            _LOGGER.debug("Resubscribing in %f seconds", wait_time)
+            if wait_time > 0:
+                await asyncio.sleep(wait_time)
+
+            await self._async_resubscribe_services(notify_errors=True)
+
+        _LOGGER.debug("_resubscribe_loop ended because of no subscriptions")
+
+    async def _update_resubscriber_task(self) -> None:
+        """Start or stop the resubscriber task, depending on having subscriptions."""
+        # Clear out done task to make later logic easier
+        if self._resubscriber_task and self._resubscriber_task.cancelled():
+            self._resubscriber_task = None
+
+        if self._subscriptions and not self._resubscriber_task:
+            _LOGGER.debug("Creating resubscribe_task")
+            # pylint: disable=fixme
+            # TODO: Use create_task instead of ensure_future with Python 3.8+
```

Comment:
Great addition, keeping python 3.6/3.7 compatibility, but already showing how to do this for newer Python versions.

Reply:
Thanks. I'm looking forward to when we can use Python 3.8 (or 3.9) features without worry.